README.markdown
NetsGraph - README
====================

+ Name: NetsGraph

+ Version: 1.0.004

+ Release date: 2012-05-10

+ Author: Nicola Asuni

+ Copyright (2012-2012):

> > Fubra Limited  
> > Manor Coach House  
> > Church Hill  
> > Aldershot  
> > Hampshire  
> > GU12 4RQ  
> > <http://www.fubra.com>  
> > <support@fubra.com>  


SOFTWARE LICENSE:
-----------------

This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License along with this program.  If not, see <http://www.gnu.org/licenses/>.

See LICENSE.TXT file for more information.


DESCRIPTION:
------------

NetsGraph is a PHP class to generate graphs (SVG and PNG format) and statistics from network data.

The class constructor accepts an array of network data and an array of options.
The data format can be: 
+ a) an array of network data arrays to be merged where the key is a time string for key and the value is a 32 bit rolling counter in bytes-per-second, or 
+ b) a simple array where the key is a UNIX EPOCH timestamp and the value is a differential value in the sampling interval expressed in bits-per-second.

The class can export data as SVG vector graph, PNG image or simple array of statistics.
Numerous options are available to change and tune the graphic output.

Two main graphs are available:
+ time: represent the Network Data Rate in a time scale.
+ percentile: represent the data ordered and calculate the specified percentile (by default 95th percentile).

The calculated statistics are:

+ samples: total number of data samples.
+ tot data: total amount of data transfer (the integral of the curve).
+ sum: sum of all samples.
+ min: minimum sample value.
+ max: maximum sample value.
+ range: difference between the maximum and minimum value on data samples.
+ mean: the arithmetic mean of the data samples values.
+ median: the numerical value separating the higher half of samples.
+ mode: the value that occurs most frequently in data samples.
+ stddev: standard deviation - shows how much variation or "dispersion" exists from the mean value.
+ skewness: measure the symmetry or asymmetry of the curve.
+ kurtosis: measure whether the data are peaked or flat relative to a normal distribution.
+ 95th prc: the value of data sample below which a certain percent (95%) of observations fall.
+ prc/max: ratio between 95th percentile and maximum value.
+ mean/prc: ratio between mean and 95th percentile.
+ med/prc: ratio between median and 95th percentile.
+ mode/prc: ratio between mode and 95th percentile.

Check the example and source code for further information.


CHANGELOG.TXT
1.0.004 (2012-05-10)
	- Units of measure were changed to conform SI standard.
	- Documentation was improved.

1.0.003 (2012-04-27)
	- Additional statistics were added.
	- Background colour was added by default.
	- PNG output was fixed.

1.0.002 (2012-04-26)
	- Display of statistics were improved.
	- Non visible point were removed to reduce file size.

1.0.001 (2012-04-25)
	- Units were added on printed statistics.
	- total data transfer was added on statistics.

1.0.000 (2012-04-24)
	- First version on GIT.

--------------------------------------------------------------------------------


netsgraph_example.php
<?php
//=============================================================================+
// File name   : netsgraph_example.php
// Begin       : 2012-04-20
// Last Update : 2012-05-10
// Version     : 1.0.004
//
// Website     : https://github.com/fubralimited/NetsGraph
//
// Description : Example file for netsgraph class.
//
// Author: Nicola Asuni
//
// (c) Copyright:
//               Fubra Limited
//               Manor Coach House
//               Church Hill
//               Aldershot
//               Hampshire
//               GU12 4RQ
//				 UK
//               http://www.rackmap.net
//               support@rackmap.net
//
// License:
//    Copyright (C) 2012-2012 Fubra Limited
//
//    This program is free software: you can redistribute it and/or modify
//    it under the terms of the GNU Affero General Public License as
//    published by the Free Software Foundation, either version 3 of the
//    License, or (at your option) any later version.
//
//    This program is distributed in the hope that it will be useful,
//    but WITHOUT ANY WARRANTY; without even the implied warranty of
//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//    GNU Affero General Public License for more details.
//
//    You should have received a copy of the GNU Affero General Public License
//    along with this program.  If not, see <http://www.gnu.org/licenses/>.
//
//    See LICENSE.TXT file for more information.
//=============================================================================+

/**
 * @file
 * Example file for netsgraph class.
 * @package com.fubra.netsgraph
 * @brief Example file for netsgraph class.
 * @author Nicola Asuni
 * @since 2012-04-20
 */

require_once('netsgraph.php');

// -----------------------------------------------------------------------------

// check input parameters

// array of raw data
//$csvdata = array('data/dataexamplehour.csv');
$csvdata = array('data/dataexampleday.csv');
//$csvdata = array('data/dataexamplemonth.csv');

// load CSV data files
$n = 0;
foreach ($csvdata as $csvfile) {
	// check if file exist
	if (file_exists($csvfile)) {
		// get CSV data into array
		$tmpdata = file($csvfile);
		if ($tmpdata[0][0] = '"') {
			// remove CSV header
			array_shift($tmpdata);
		}
		foreach ($tmpdata as $row) {
			// separate data
			$col = explode(',', $row);
			$data[$n][trim($col[0])] = $col[1];
		}
		++$n;
	}
}

// set parameters
$p = array();
$p['graph_type'] = 'percentile'; // 'time' or 'percentile'
//$p['label_x'] = 'Percentile [%]';
//$p['label_y'] = 'Network Data Rate [Mbps]';
$p['graph_width'] = 800;
//$p['graph_height'] = 450;
$p['background_color'] = '#ffffff';
$p['border_width'] = 1;
$p['border_color'] = '#000000';
$p['print_stats'] = true;
$p['stats_unit_prefix'] = true;
$p['stats_bgcolor'] = '#ffffee';
$p['stats_fontfamily'] = 'monospace';
$p['stats_fontweight'] = 'normal';
$p['stats_fontsize'] = 10;
$p['stats_fontcolor'] = '#404040';
$p['gridy_fontfamily'] = 'Arial,Verdana';
$p['gridy_fontweight'] = 'normal';
$p['gridy_fontsize'] = 12;
$p['gridy_fontcolor'] = '#808080';
$p['gridx_fontfamily'] = 'Arial,Verdana';
$p['gridx_fontweight'] = 'normal';
$p['gridx_fontsize'] = 12;
$p['gridx_fontcolor'] = '#808080';
$p['grid_vert_line_width'] = 1;
$p['grid_vert_line_color'] = '#cccccc';
$p['grid_horiz_line_width'] = 1;
$p['grid_horiz_line_color']= '#cccccc';
$p['label_x_fontfamily'] = 'Arial,Verdana';
$p['label_x_fontweight'] = 'normal';
$p['label_x_fontsize'] = 14;
$p['label_x_fontcolor'] = '#000000';
$p['label_y_fontfamily'] = 'Arial,Verdana';
$p['label_y_fontweight'] = 'normal';
$p['label_y_fontsize'] = 14;
$p['label_y_fontcolor'] = '#000000';
$p['data_line_color'] = '#0000ff';
$p['data_fill_color'] = '#00ff00';
$p['data_fill_opacity'] = 0.7;
$p['data_line_width'] = 1;
$p['percentile_x'] = 95;
$p['percentile_horiz_line_width'] = 1;
$p['percentile_horiz_line_color'] = '#ff0000';
$p['percentile_horiz_line_opacity'] = 1;
$p['percentile_vert_line_width'] = 1;
$p['percentile_vert_line_color'] = '#ff0000';
$p['percentile_vert_line_opacity'] = 1;
$p['percentile_gridy_fontfamily'] = 'Arial,Verdana';
$p['percentile_gridy_fontweight'] = 'bold';
$p['percentile_gridy_fontsize'] = 14;
$p['percentile_gridy_fontcolor'] = '#b00000';
$p['percentile_gridx_fontfamily'] = 'Arial,Verdana';
$p['percentile_gridx_fontweight'] = 'bold';
$p['percentile_gridx_fontsize'] = 12;
$p['percentile_gridx_fontcolor'] = '#ff0000';

// generate SVG
$svg = new NetsGraph($data, $p);

// output SVG file (including headers)
$svg->getSVG();

// output PNG file (including headers)
//$svg->getPNG();

//=============================================================================+
// END OF FILE
//=============================================================================+
